src/main.rs

```diff
@@ -34,321 +33,323 @@ extern crate xi_core_lib;
+        let core = Arc::downgrade(&self.core);
+        let state = self.state.clone();
+        self.core.lock().unwrap().send_request("new_view", &params,
             move |value| {
-                let value = value.clone();
-                handle.add_idle(move |a| {
-                    let handler = a.downcast_ref::<MainWinHandler>().unwrap();
-                    let edit_view = &mut handler.win.state.borrow_mut().edit_view;
-                    edit_view.set_view_id(value.as_str().unwrap());
-                });
+                let view_id = value.clone().as_str().unwrap().to_string();
+                let mut state = state.lock().unwrap();
+                let handle = handle.clone();
+                state.focused = view_id.clone();
+                state.views.insert(view_id.clone(),
+                    ViewState {
+                        id: 0,
+                        filename: filename.clone(),
+                        handle: handle.clone(),
+                    }
+                );
+                UiMain::send_ext(&handle.lock().unwrap(), edit_view, EditViewCommands::Core(core));
+                UiMain::send_ext(&handle.lock().unwrap(), edit_view, EditViewCommands::ViewId(view_id));
             }
         );
     }
 
     fn handle_cmd(&self, method: &str, params: &Value) {
-        let mut state = self.state.borrow_mut();
         match method {
-            "update" => state.edit_view.apply_update(&params["update"]),
-            "scroll_to" => state.edit_view.scroll_to(params["line"].as_u64().unwrap() as usize),
+            "update" => self.send_view_cmd(EditViewCommands::ApplyUpdate(params["update"].clone())),
+            "scroll_to" => self.send_view_cmd(EditViewCommands::ScrollTo(params["line"].as_u64().unwrap() as usize)),
             "available_themes" => (), // TODO
             "available_plugins" => (), // TODO
+            "available_languages" => (), // TODO
             "config_changed" => (), // TODO
+            "language_changed" => (), // TODO
             _ => println!("unhandled core->fe method {}", method),
         }
-        // TODO: edit view should probably handle this logic
-        self.invalidate();
-    }
-
-    pub fn invalidate(&self) {
-        self.handle.borrow().invalidate();
     }
 }
 
-fn create_main(core: Core) -> Result<WindowHandle, Error> {
-    let main_state = MainWinState::new();
-    let main_win = Rc::new(MainWin::new(core, main_state));
-    let main_win_handler = MainWinHandler {
-        win: main_win,
-    };
-
-    let menubar = menus::create_menus();
-
-    let mut builder = WindowBuilder::new();
-    builder.set_handler(Box::new(main_win_handler));
-    builder.set_title("xi-editor");
-    builder.set_cursor(Cursor::IBeam);
-    builder.set_menu(menubar);
-    let window = builder.build().unwrap();
-    Ok(window)
-}
-
 #[derive(Clone)]
-struct MyHandler {
-    runloop: RunLoopHandle,
-    win_handle: Arc<Mutex<Option<IdleHandle>>>,
+struct AppDispatcher {
```

Comment:
Again it doesn't seem necessary that there's an additional `Arc<Mutex<>>` around this, but I'm less certain you can get rid of it.

Reply:
I couldn't get rid of it successfully. (It doesn't give any errors, but it won't work.) 🤷‍♂️